hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/JobRepository.java

```diff
@@ -346,10 +361,17 @@ void cleanup(NodeEngine nodeEngine) {
         int maxNoResults = Math.max(1, nodeEngine.getProperties().getInteger(JetProperties.JOB_RESULTS_MAX_SIZE));
         // delete oldest job results
         if (jobResults.size() > Util.addClamped(maxNoResults, maxNoResults / MAX_NO_RESULTS_OVERHEAD)) {
-            jobResults.values().stream().sorted(Comparator.comparing(JobResult::getCompletionTime).reversed())
+            jobResults.values().stream().sorted(comparing(JobResult::getCompletionTime).reversed())
                       .skip(maxNoResults)
-                      .forEach(r -> jobResults.remove(r.getJobId()));
+                      .map(JobResult::getJobId)
+                      .collect(Collectors.toSet())
+                      .forEach(id -> {
+                          jobResults.delete(id);
+                          jobMetrics.delete(id);
```

Comment:
We should first delete metrics and then results. Currently if the member fails in between these two, we can have result deleted but not metrics and the metrics will leak. With reverse order, we'll find the result again, just the metrics will be gone.